tests/utils/test_nbis_xml_generator.py

```diff
@@ -0,0 +1,176 @@
+        
+        self.outdir = tempfile.mkdtemp()
+        self.xgen = xml_generator(self.pid, outdir=self.outdir, LOG=self.log, pcon=self.pcon, fcon=self.fcon, xcon=self.xcon)
+        
+        self.stats = {
+            'experiment': 
+                {'selection': 'unspecified', 
+                 'protocol': 'NA', 
+                 'library': 'P12345_1001_prep', 
+                 'design': 'Sample library for sequencing on Illumina MiSeq', 
+                 'discriptor': 'P12345_1001', 
+                 'layout': '<PAIRED></PAIRED>', 
+                 'source': 'METAGENOMIC', 
+                 'study': 'P12345', 
+                 'title': 'Prep for P12345_1001 sequenced in Illumina MiSeq',
+                 'strategy': 'OTHER', 
+                 'alias': 'P12345_1001_A_illumina_miseq_experiment',
+                 'instrument': 'Illumina MiSeq'
+                 },
+                'run': 
+                    {'exp_ref': 'P12345_1001_A_illumina_miseq_experiment', 
+                     'alias': 'P12345_1001_A_illumina_miseq_runs', 
+                     'data_name': 'P12345_1001', 
+                     'files': '\t\t\t\t<FILE filename="file1.fastq.gz" filetype="fastq" checksum_method="MD5" checksum="sum" />\n'
+                     }
+                    }
+    
+    @classmethod
+    def tearDownClass(self):
+        shutil.rmtree(self.outdir)
+    
+    @patch('taca_ngi_pipeline.utils.nbis_xml_generator.xml_generator._collect_sample_stats')
+    @patch('taca_ngi_pipeline.utils.nbis_xml_generator.xml_generator._generate_manifest_file')
+    def test_generate_xml_and_manifest(self, mock_generate, mock_collect):
+        mock_collect.return_value = [self.stats]
+        self.xgen.generate_xml_and_manifest()
+        got_exml = os.path.join(self.outdir, 'P12345_experiments.xml')
+        exp_exml = os.path.join('data', 'P12345_experiments.xml')
+        got_rxml = os.path.join(self.outdir, 'P12345_runs.xml')
+        exp_rxml = os.path.join('data', 'P12345_runs.xml')
+ 
+        self.assertTrue(filecmp.cmp(got_exml, exp_exml))
+        self.assertTrue(filecmp.cmp(got_rxml, exp_rxml))
+    
+    def test__generate_manifest_file(self):
+        experiment_details = {'study': 'P12345',
+                       'discriptor': 'P12345_1001',
+                       'alias': 'alias',
+                       'instrument': 'inst',
+                       'source': 'src',
+                       'selection': 'sel',
+                       'strategy': 'strat',
+                       'layout': '<PAIRED></PAIRED>'}
+        self.xgen._generate_manifest_file(experiment_details, 'run_details')
+        got_manifest_path = os.path.join(self.outdir, 'manifestFiles', 'P12345_1001_manifest.txt')
+        self.assertTrue(filecmp.cmp(got_manifest_path,'data/P12345_1001_manifest.txt'))
+    
+    @patch('taca_ngi_pipeline.utils.nbis_xml_generator.xml_generator._generate_files_block')
+    def test__collect_sample_stats(self, mock_blockgen):
```

Comment:
Two '_' in the function name here and in more names in the following code

Reply:
Yup, that's because the function is called "_collect_sample_stats". I'm not sure what's best practice in that case, I've seen both versions. What do you think?

Reply:
Ah, I see. Maybe a docstring with the name of the function it tests could clarify things? It looks ok as it is though.

Reply:
Good idea! 👍 I was lazy and skipped them this time 😉 I'll revisit the tests to try to improve the coverage further once I'm back, so I can add them then.

Reply:
Good idea! 👍 I was lazy and skipped them this time 😉 I'll revisit the tests to try to improve the coverage further once I'm back, so I can add them then.